### ISSUE TITLE
Change log / Release

### ISSUE THREAD

**Issue by vince62s** [kaldi-asr/kaldi#1943]
This is just for discussion but feel free to close it of not relevant.

I am reacting to this new issue about backstich "best practice" as well as 2 new recent features that were commited: attention modelling layer #1731 and non splicing layers #1937
These are majors features IMO that are put in the middle of hundreds of commmits on this page: http://kaldi-asr.org/doc/versions.html where commits "fix typo" are at the same level as major features.

My suggestion, would be to use the "release" tab of github to at least list all major new features for each version.
This is not a big maintenance, and would enable users to see at first sight what is really important when the repo is moving so fast....(for the good).

Cheers.

**Comment by danpovey** [comment]
It's not clear to me that the releases feature of github is really
compatible with our workflow and other practices.  But I'll try to think
about the best way to give more visibility to important changes.  Any
concrete suggestions would be welcomed.

**Comment by vince62s** [comment]
My understanding is that this is just tags. You decide which commit is a version (major or minor but in Kaldi's case for sure not all minors) even starting with 5.1 and 5.2 and main major features would be great.
Here is an example: https://github.com/tensorflow/tensor2tensor/releases
But, yes this is manual.

**Comment by kkm000** [comment]
We actually now have tags for 5.1 through 5.4. But 5.5 was not tagged. I'll take care of this, thanks. Since there appears to me a mash of commits after the 5.5 branch has been merged, I'll track this in a new issue.

**Comment by danpovey** [comment]
The tags represent the *end* of development of the respective branches.
Because 5.5 is the current version, it corresponds to master, we'd only
create the tag once we go to 5.6.

**Comment by kkm000** [comment]
This is a bit more complex with 5.5. A stable 5.5 is essentially the end of the branch *plus* 2 bug fixes that came in the following ~2 weeks. One is a memory leak, another broken compile on Windows. I'll explain in a ticket. I need to find them first.

**Comment by kkm000** [comment]
Oh. I am nor really sure now. I've seen a huge branch merge at c0b18b77b on 2018-09-03, after which the current version was changed to 5.5. Our tools report 5.5, since that is in the version file, and my understanding was this is what indicates the current version. I am talking about this graph (the tag 'apparent-5.5' I added locally):

```
* 1669d2401 [egs,scripts] chime-4 advanced baseline (#2142)
*   a0bc18edc [src] Remove kaldi-gpsr.{h,cc} which was not used.
|\
| *   c0b18b77b resolved conflicts
| |\
| |/
|/|
* |   f0a793112 [src] Fix remaining warnings caused by #2411
|\ \
| * | 9f0db6374 (tag: apparent-5.5) Two more small fixes.
| * | 8d8c5af55 Fix style errors.
| * | 0621b792b Fix remaining -Wmaybe-unitialized warnings.
|/ /
* |   bda1dc7cf [src,scripts,egs] Grammar decoding; upgrade version to 5.5.
|\ \
| * | bdcdd4722 [doc] Update version documentation for version 5.5.
| * | 7aab92b7c [build] Upgrade version to 5.5
| * | 8b0bc2b89 [src] Fix leak in make-grammar-fst
| * | 074a1d9eb [src,egs] Cosmetic changes, mostly fixes to comments.
| * | 5cd7cde59 [scripts] Small fix in grammar-decoding script
| * | 03b585468 [doc] update grammar-fst documentation
| * | d0c68a60e [src] Refactor online decoder; get grammar decoding work in online case.
| * | a39b15c2b [egs] Small example script fixes
```

**Comment by kkm000** [comment]
So it looked like there was a big merge before the version was changed, as if it was waiting until the trunk version stabilizes. But there are 2 essential fixes made to it later.

**Comment by danpovey** [comment]
I am OK to take those fixes, merge them into the 5.4 branch, and move the tag forward.

**Comment by kkm000** [comment]
But what does the version signify? What does this change in the master branch mean: the development of the version 5.5 *started* from this point on, or was this the *release* of 5.5?
```diff
$ git show 7aab92b7c
commit 7aab92b7c2cca88875a92312dd4793c737ad7f03
Author: Daniel Povey <dpovey@gmail.com>
Date:   Sun Sep 2 13:15:35 2018 -0400

    [build] Upgrade version to 5.5

diff --git a/src/.version b/src/.version
index 37c2d9960..9ad974f61 100644
--- a/src/.version
+++ b/src/.version
@@ -1 +1 @@
-5.4
+5.5
```

**Comment by danpovey** [comment]
That is the start of branch 5.5: would be equivalent to 5.5.0.

the tag 5.4 indicates the end of branch 5.4, e.g. 5.4.big-number

**Comment by kkm000** [comment]
Well, we have no tags, only branches. There is a 5.4 *branch*, but it has not advanced off master (has no commits not in master, directly reachable from the head of master).

So we essentially *do* have a release 5.5, right?

**Comment by danpovey** [comment]
OK, branches then, not tags.  As far as I am concerned there is no concept
of releases.  If you want you can apply fixes to the older branches like
5.4.  That's why I created those branches-- to record the last
working/compatible state of those version numbers and allow the opportunity
for small patches to be added to them.

**Comment by kkm000** [comment]
Ok. Now I'm lost thoroughly, but I do not think this is important, really. I'll leave it as it is.

I think I also misunderstood the @vince62s's concern. So the doc page essentially just a git log, and the users would rather see a list of major functional changes only. I think I'll leave the issue open then, but I do not really know how to even approach this.

**Comment by vince62s** [comment]
@kkm000 My comment is quite old and Dan obviously did not want to maintain the "release concept" in github.
Another example is here
https://github.com/OpenNMT/OpenNMT-py/releases
It's just a link between an arbitrary ralease number and a tag. Same concept as Dan's but in Github's world.

**Comment by danpovey** [comment]
I don't want to go that route because it might steer people towards using
older versions of Kaldi, which would increase maintenance workload, since
we would have to keep track of which bugs exist in older versions.

**Comment by kkm000** [comment]
I think the main concern here is that the version log http://kaldi-asr.org/doc/versions.html is too fine-grained. For example, this is a "patch release" version:

> 5.5.262 6134c290f 2019-03-20 [scripts] Fix bug in comment (#3152)

and the changeset, in its entirety,

```diff
index 34476fdb3..335e69e9a 100755
--- a/egs/wsj/s5/utils/parse_options.sh
+++ b/egs/wsj/s5/utils/parse_options.sh
@@ -44,3 +44,3 @@ done
 ###
-### No we process the command line options
+### Now we process the command line options
 ###
```

So marking important changes does make sense, but I really have no idea how to define "important" in this sense, and how not to forget to maintain that. The latter part is the trickiest one. : )

**Comment by danpovey** [comment]
We could add a tag [minor] or something like that, to suppress unimportant
changes from appearing there.

**Comment by danpovey** [comment]
Or better, a tag [major], e.g. [src,egs,major] ... since most PRs are minor.

**Comment by kkm000** [comment]
This, I believe, could be automatically tagged. I'll check what tools/bots are available for GitHub for this.

**Comment by danpovey** [comment]
I'd rather do it personally, since what is major or not does not have to do
with how many lines of code changed.

**Comment by kkm000** [comment]
Right, this is of course manual. What I mean is a tool/script/thing which, upon seeing the [..major..] marker, slaps a tag on it, and lists a changelog to the previous tag/marker in the release notes on GitHub.

**Comment by danpovey** [comment]
I am not planning to use the GitHub release mechanism, because I think it
goes against the dynamics of the project.

**Comment by kkm000** [comment]
No, not full-blown "releases", like with the source tarball. That would not make sense, I'm totally with you. Simply tagging and adding a tag message.

But then, I do not know if this is really worth the effort. There is a lot of changes that are important, even if one-line bug fixes. I samples the 20 latest commits, and marked '(!)' those that are likely "important": either bug fixes or new features; I counted 12. It's really unsimple to decide what is [major], easier said than done. If only mark a feature, then the initial commit is usually followed by bug fixes anyway. Then the `[major]` becomes rather confusing.

Maybe we should change exactly nothing? :)

 1. (!) [scripts] Fix bug in extend_lang.sh regarding extra_disambig.txt (#3195)
 1. (!) [build] Add missing dependency to Makefile (#3191)
 1. [github] Add GitHub issue templates (#3187)
 1. [src] Efficiency improvement and extra checking for cudamarix, RE def…
 1. (!) [scripts] Fix non-randomness in getting utt2uniq, introduced in #3142 (…
 1. (!) [build] Add new nvidia tools to windows build (#3159)
 1. [src] Disable unget warning in PeekToken (and other small fix) (#3163)
 1. (!) [scripts] Fix bug in steps/segmentation/ali_to_targets.sh (#3155)
 1. (!) [src] Add binary that functions as a TCP server (#2938)
 1. [src] Fix typo in comment (#3147)
 1. [build] Make sure PaUtils exported from portaudio (#3144)
 1. (!) [src] Fix to "Fixes to grammar-fst & LM-disambig symbols" (#3000) (#3143
 1. (!) [src] Fix bad assert in fstmakecontextsyms (#3142)
 1. (!) [scripts] Bug-fix for removing deleted words (#3116)
 1. (!) [egs] Add "formosa_speech" recipe (Taiwanese Mandarin ASR) (#2474)
 1. [scripts] Simplify text encoding in RNNLM scripts (now only support u…
 1. [doc] Small documentation fixes; update on Kaldi history (#3031)
 1. [src,scripts] Python2 compatibility fixes and code cleanup for nnet1 (#…
 1. (!) [src] Fix wrong assertion failure in nnet3-am-compute (#3106)
 1. (!) [scripts] Fix frame_shift bug in egs/swbd/s5c/local/score_sclite_conf…

**Comment by danpovey** [comment]
Yeah, I think the status quo is OK.

**Comment by darkdragon-001** [comment]
Having some sort of git tags (even without changelog and backporting fixes) would make it easier to package this for Linux distros like Fedora or Debian.
